### PR TITLE
[Feature] different number of actions for each action dimension (discrete actions)

### PR DESCRIPTION
**Pull request by rikifunt**
This PR supports specifying different number of actions for each action dimension when the action space is discrete. The motivation is to support action dimensions with different semantics that require different discretizations for each dimension (e.g. combining navigation and "boolean" actions). More in general, this should allow to add any kind of discrete action with arbitrary semantics (also thanks to process_action).

To implement this, I added an `action_nvec` property to `Dynamics`, which is "mutually exclusive" with `needed_action_size`, meaning that if one is overridden the other is computed automatically, and vice-versa (they can't be both overridden). By default, `action_nvec` will be a list of 3s of length `action_size` when not specified, to keep compatibility with existing scenarios. I also added `action_nvec` to the `Agent` class, which is then used in `Environment` to build the discrete action space(s) and in `Environment._set_action` to: 1) convert discrete actions in the interval [0, n) to [-u_max, u_max]; and 2) convert discrete actions to multi-discrete ones when needed.

I also added a `Composite` subclass of `Dynamics` that allows to compose multiple `Dynamics` into one by concatenating their actions, which should make it more convenient (modular) to define complex actions/dynamics.

Let me know if there is anything else that can improve the PR.

### CODE REVIEW

File: vmas/simulator/environment/environment.py [hunk 1]
```diff
@@ -490,41 +513,40 @@ def _set_action(self, action, agent):
...
+                nvec = list(agent.action_nvec) + (
+                    [self.world.dim_c]
+                    if not agent.silent and self.world.dim_c != 0
+                    else []
                 )
-                action = action_range.nonzero()[:, 1:]
+                for n in nvec:
+                    actions.append(flat_action % n)
+                    flat_action = flat_action // n
+                action = torch.stack(actions, dim=-1)
```
Comment (matteobettini) [inline vmas/simulator/environment/environment.py:532]: This changes the logic: before we had 0-> [0,0,0], 1-> [0,0,1], 2-> [0,0,2], 3-> [0,1,0], 4-> [0,1,1] and so on

File: vmas/simulator/environment/environment.py [hunk 2]
```diff
@@ -490,41 +513,40 @@ def _set_action(self, action, agent):
...
-                disc_action_value = agent.action.u_range_tensor[action_index]
-                agent.action.u[:, action_index] -= disc_action_value * arr1.squeeze(-1)
-                agent.action.u[:, action_index] += disc_action_value * arr2.squeeze(-1)
+                u_max = agent.action.u_range_tensor[action_index]
+                # We know u must be in [-u_max, u_max], and we know action is
+                # in [0, n-1]. Conversion steps: [0, n-1] -> [0, 1] -> [0, 2*u_max] -> [-u_max, u_max]
+                # E.g. action 0 -> -u_max, action n-1 -> u_max, action 1 -> -u_max + 2*u_max/(n-1)
+                agent.action.u[:, action_index] = (physical_action / (n - 1)) * (
+                    2 * u_max
+                ) - u_max
```
Comment (matteobettini) [inline vmas/simulator/environment/environment.py:549]: This also changes the logic as before we had action 0 always being no-op followed by a -u_range action and a +u_range action.

**Review by matteobettini — COMMENTED** [review]
Wow thanks for this, it is really well executed!

it is something i also have been thinking about a lo

I think we definitely need to allow different discretizations other than 3, but we have to be careful about bc-compatibility here.

Also another point is that I would like to keep the core and dynamic agnostic of wether the input was discrete or continuous. This might not be possible for the agent (as we need to somehow tell how we want to discretize actions) but it is possible with dynamics. In particular:

- I do not think the dynamics need to know how many discrete actions the action is divided into (and even if they do they can access it via the agent). So I would have the nvec property just as part of the agent and not also of the dynamics
- The way the dynamics are designed is related to motion, so it does not make sense to have more than one. This is why HolonomicWithRotation exists and it is not Holonomic+Rotation. If you need to preprocess certain actions (e.g., for a gripper) I would create a custom component that you call in `process_action` which consumes the actions it needs leaving the ones that will later go to the dynamics

To recap, I would:
- remove the changes to the dynamics (we can keep the rotation only dynamics in its own file)
- Think a bit more about the bc-comp implications of the changes in the env, probably `interactive_render` has to be adjusted

We will also need some tests for the env changes

**Comment by rikifunt** [comment]
Thank you for the comments and clarifications! I will address them and continue working on this as soon as I can.

It makes a lot of sense to keep the dynamics agnostic of the action space. To keep compatibility with the current action spaces, it should be possible to rewrite the multidiscrete -> continuous conversion in `Environment._set_action` so that the discretization logic stays the same for n=3, I'll look into it.

At this point, the overall changes would be:
- adding the `action_nvec` parameter to `Agent.__init__` (optional and mutually exclusive with `action_size`)
- building discrete spaces in `Environment` and sampling random actions according to `action_nvec`
- using `action_nvec` in `Environment._set_action`, keeping the current logic when n=3
- adding tests where needed
Then, as you said, we can delegate any extra action semantics to process_action. Are these ok?

I will also look into interactive_render (I initially left it untouched because it uses continuous actions). Let me know if there is other functionality that you think these changes might break.

**Comment by matteobettini** [comment]
> Thank you for the comments and clarifications! I will address them and continue working on this as soon as I can.
> 
> It makes a lot of sense to keep the dynamics agnostic of the action space. To keep compatibility with the current action spaces, it should be possible to rewrite the multidiscrete -> continuous conversion in `Environment._set_action` so that the discretization logic stays the same for n=3, I'll look into it.
> 
> At this point, the overall changes would be:
> 
> * adding the `action_nvec` parameter to `Agent.__init__` (optional and mutually exclusive with `action_size`)

Yes, you basically have already done this. I would not make them mutually exclusive. They will default to `None`. If one is specified, we create the other one based on that one. If both are specified, we check they are coherent. If none are specified we default to dynamics (and the old 3 way split)

Also maybe it is worth naming this with something that contains "discrete" like `discrete_action_nvec` to highlight that is not something that interests users that have continuous actions (which I think is majority)

> * building discrete spaces in `Environment` and sampling random actions according to `action_nvec`

yes, merge main and this should come almost for free with what you have already done

> * using `action_nvec` in `Environment._set_action`, keeping the current logic when n=3

so here is where i am still a bit undecided.
what I suggest is that we split the range among the n actions. 
When `n` is odd, we make the first action being 0 for bc-compatibility.
This would make so that n=3 is not a special case but the way we handle odd `n`s

 For example:
- `n=2` leads to `[-range,range]`
- `n=3` leads to `[0,-range,range]`
- `n=4` leads to `[-range,-range/3,range/3,range]`
- `n=5` leads to `[0,-range,-range/2,range/2,range]`


> * adding tests where needed

Yeah thanks. the main tests I would like to see is regarding the logic that will be added to `environemnt._set_action` in different settings

>   Then, as you said, we can delegate any extra action semantics to process_action. Are these ok?

yep!

> I will also look into interactive_render (I initially left it untouched because it uses continuous actions). Let me know if there is other functionality that you think these changes might break.

Oh yes sorry I forgot interactive_render uses continuous, then we should be good there

**Comment by rikifunt** [comment]
I added 2 commits with the suggested changes:
- `action_nvec` is now only present in `Agent` and is called `discrete_action_nvec` (changes to dynamics are reverted)
- `discrete_action_nvec` is not mutually exclusive with `action_size`, but we check that they are consistent in `Agent.__init__`
- the logic in `_set_action` handles odd `n`s by converting the first discrete action to `u = 0`, and shifting the remaining "negative" actions by -1 so that the previous logic applies
- the comment in the discrete to multidiscrete conversion in `_set_action` is now accurate (turns out the logic was already the same as before the generalization)

Overall, this results in the following "multidiscrete `a`" to "continuous `u`" conversions (where `U = u_max`):
- when `n` is odd (and let `h = (n-1)/2`, i.e. the middle point of the discrete possibilities):
  ```
  a == 0 ==> u == 0
  a in [1, h] ==> u in [-U, -U/h]
  a in [h+1, n) ==> u in [+U/h, U]
  ```
- when `n` is even
  ```
  a in [0, n] ==> u in [-U, U]
  ```

I am now working on adding tests that check the above logic and the discrete-multidiscrete conversion, and I was wondering if it is possible to build a `Scenario` object directly instead of passing a string to `make_env`? This would make it possible to set the `nvec`s to values other than 3 before constructing the `Environment` object, so that various values can be tested. Currently I am trying to do something like this in `tests/test_vmas.py` (where the `load(scenario)` line is taken from `make_env`):
```python
from vmas import scenarios

...

@pytest.mark.parametrize("scenario", scenario_names())
def test_discrete_action_nvec_multidiscrete(scenario, num_envs=10, n_steps=10):
    scenario = scenarios.load(scenario).Scenario()
    random.seed(0)
    for agent in scenario.agents:
        agent.discrete_action_nvec = [random.randint(2, 6) for _ in range(agent.action_size)]
    env = make_env(
        scenario=scenario,
        num_envs=num_envs,
        seed=0,
        multidiscrete_actions=True,
        continuous_actions=False,
    )
    for _ in range(n_steps):
        ...
```
However, I get an error on the first line of the function, saying that `scenarios` is a `list` and does not have the `load` method. I suspect this has something to do with the python import machinery? I also tried using `vmas.scenarios` directly instead of importing it at the top, with the same result. From what I can see, `vmas.scenarios` is both a subpackage in the `vmas` directory and a list defined in `vmas/__init__.py`, so maybe this is the problem? Is there any other way to accomplish this?

File: vmas/simulator/core.py [hunk 3]
```diff
@@ -862,6 +862,9 @@ def __init__(
         render_action: bool = False,
         dynamics: Dynamics = None,  # Defaults to holonomic
         action_size: int = None,  # Defaults to what required by the dynamics
+        discrete_action_nvec: List[
+            int
+        ] = None,  # Defaults to 3-way discretization (stay, decrement, increment)
```
Comment (matteobettini) [inline vmas/simulator/core.py:867]:
```suggestion
        ] = None,  # Defaults to 3-way discretization if discrete actions are chosen (stay, decrement, increment)
```

**Comment by matteobettini** [comment]
Fantastic work! 

It seems to be exactly what we want

> I am now working on adding tests that check the above logic and the discrete-multidiscrete conversion, and I was wondering if it is possible to build a `Scenario` object directly instead of passing a string to `make_env`? This would make it possible to set the `nvec`s to values other than 3 before constructing the `Environment` object, so that various values can be tested. Currently I am trying to do something like this in `tests/test_vmas.py` (where the `load(scenario)` line is taken from `make_env`):
> 
> ```python
> from vmas import scenarios
> 
> ...
> 
> @pytest.mark.parametrize("scenario", scenario_names())
> def test_discrete_action_nvec_multidiscrete(scenario, num_envs=10, n_steps=10):
>     scenario = scenarios.load(scenario).Scenario()
>     random.seed(0)
>     for agent in scenario.agents:
>         agent.discrete_action_nvec = [random.randint(2, 6) for _ in range(agent.action_size)]
>     env = make_env(
>         scenario=scenario,
>         num_envs=num_envs,
>         seed=0,
>         multidiscrete_actions=True,
>         continuous_actions=False,
>     )
>     for _ in range(n_steps):
>         ...
> ```
> 
> However, I get an error on the first line of the function, saying that `scenarios` is a `list` and does not have the `load` method. I suspect this has something to do with the python import machinery? I also tried using `vmas.scenarios` directly instead of importing it at the top, with the same result. From what I can see, `vmas.scenarios` is both a subpackage in the `vmas` directory and a list defined in `vmas/__init__.py`, so maybe this is the problem? Is there any other way to accomplish this?

if you do `from vmas import scenarios` it will load the scenario list. 

What you want is the module:

```
@pytest.mark.parametrize("scenario", scenario_names())
def test_prova(scenario, num_envs=10, n_steps=10):
    from vmas.scenarios import load

    scenario_class = load(f"{scenario}.py").Scenario()
    env = make_env(
        scenario=scenario_class,
        num_envs=num_envs,
        seed=0,
        multidiscrete_actions=True,
        continuous_actions=False,
    )
    for _ in range(n_steps):
        env.step(env.get_random_actions())
```
this will work

**Review by matteobettini — COMMENTED** [review]
Amazing!

For the tests let's make sure we test:

- `discrete` and `multidiscrete` action spaces
-  different nvecs with all the edge cases
-  the sampling logic

but judging from the work so far i am confident the tests are gonna be strong

File: vmas/simulator/core.py [hunk 4]
```diff
@@ -914,9 +923,14 @@ def __init__(
...
-            action_size if action_size is not None else self.dynamics.needed_action_size
-        )
+        if action_size is not None:
+            self.action_size = action_size
+        elif discrete_action_nvec is not None:
+            self.action_size = len(discrete_action_nvec)
+        else:
+            self.action_size = self.dynamics.needed_action_size
+        if discrete_action_nvec is None:
+            self.discrete_action_nvec = [3] * self.action_size
```
Comment (matteobettini) [inline vmas/simulator/core.py:933]:
```suggestion
        if discrete_action_nvec is None:
            self.discrete_action_nvec = [3] * self.action_size
        else:
           self.discrete_action_nvec = discrete_action_nvec
```

**Comment by matteobettini** [comment]
`n=1` is not currently handled properly and results in NaNs

File: vmas/simulator/environment/environment.py [hunk 5]
```diff
@@ -503,41 +504,48 @@ def _set_action(self, action, agent):
...
-                disc_action_value = agent.action.u_range_tensor[action_index]
-                agent.action.u[:, action_index] -= disc_action_value * arr1.squeeze(-1)
-                agent.action.u[:, action_index] += disc_action_value * arr2.squeeze(-1)
+                u_max = agent.action.u_range_tensor[action_index]
+                # For odd n we want the first action to always map to u=0, so
+                # we swap 0 values with the middle value, and shift the first
+                # half of the remaining values by -1.
+                if n % 2 != 0:
+                    stay = physical_action == 0
+                    decrement = (physical_action > 0) & (physical_action < n // 2)
```
Comment (matteobettini) [inline vmas/simulator/environment/environment.py:540]:
```suggestion
                    decrement = (physical_action > 0) & (physical_action <= n // 2)
```

File: vmas/simulator/environment/environment.py [hunk 6]
```diff
@@ -503,41 +504,48 @@ def _set_action(self, action, agent):
...
+                nvec = list(agent.discrete_action_nvec) + (
+                    [self.world.dim_c]
+                    if not agent.silent and self.world.dim_c != 0
+                    else []
                 )
-                action = action_range.nonzero()[:, 1:]
+                for n in nvec:
+                    actions.append(flat_action % n)
+                    flat_action = flat_action // n
+                action = torch.stack(actions, dim=-1)
```
Comment (matteobettini) [inline vmas/simulator/environment/environment.py:523]: There also seems to be a little inconsistency with this bit

with `nvec=[3,3]` and `action=tensor([7, 6, 3, 0, 1, 7, 5, 6, 3, 7, 4, 4, 7, 7, 1, 3, 2, 3, 6, 3, 0, 6, 8, 0,
        3, 2, 2, 3, 4, 5, 8, 8])`

i am getting

## PR

```
tensor([[1, 2],
        [0, 2],
        [0, 1],
        [0, 0],
        [1, 0],
        [1, 2],
        [2, 1],
        [0, 2],
        [0, 1],
        [1, 2],
        [1, 1],
        [1, 1],
        [1, 2],
        [1, 2],
        [1, 0],
        [0, 1],
        [2, 0],
        [0, 1],
        [0, 2],
        [0, 1],
        [0, 0],
        [0, 2],
        [2, 2],
        [0, 0],
        [0, 1],
        [2, 0],
        [2, 0],
        [0, 1],
        [1, 1],
        [2, 1],
        [2, 2],
        [2, 2]])
```

## vmas main

```
tensor([[2, 1],
        [2, 0],
        [1, 0],
        [0, 0],
        [0, 1],
        [2, 1],
        [1, 2],
        [2, 0],
        [1, 0],
        [2, 1],
        [1, 1],
        [1, 1],
        [2, 1],
        [2, 1],
        [0, 1],
        [1, 0],
        [0, 2],
        [1, 0],
        [2, 0],
        [1, 0],
        [0, 0],
        [2, 0],
        [2, 2],
        [0, 0],
        [1, 0],
        [0, 2],
        [0, 2],
        [1, 0],
        [1, 1],
        [1, 2],
        [2, 2],
        [2, 2]])

```

Comment (matteobettini) [reply vmas/simulator/environment/environment.py:523]: `action = torch.stack(actions, dim=-1).flip(-1)` this should fix it

Comment (rikifunt) [reply vmas/simulator/environment/environment.py]: Thanks for catching this! I fixed it and added relevant tests

**Comment by rikifunt** [comment]
Thanks for the comments! I've committed the suggested changes and fixed the bug in the discrete -> multi-discrete conversion

> `n=1` is not currently handled properly and results in NaNs
> 
> should this even be allowed? I would introduce a check that reads the nvec and checks all values are >= 2

Yes, this should not happen, I added an input sanity check in `Agent.__init__`.

To make sure all of this works as intended, I've added some tests that check:
- if randomly sampled actions are in the computed action space
- if multi-discrete actions map to the expected continuous control after `Environment._set_action`
- if discrete actions that come from the product of multi-discrete spaces are converted properly to multi-discrete actions in `Environment._set_action`

These are run for all scenarios that don't explicitly override `process_action` (since it breaks the action <-> control mapping). `nvec`s are randomly sampled (with a fixed seed) in the interval [2,6] to properly test mixed odd-even `nvec`s and odd n != 3. Let me know if there is anything else that needs testing.

File: tests/test_vmas.py [hunk 7]
```diff
@@ -70,6 +72,135 @@ def test_multi_discrete_actions(scenario, num_envs=10, n_steps=10):
         env.step(env.get_random_actions())
 
 
+@pytest.mark.parametrize("scenario", scenario_names())
+def test_discrete_action_nvec(scenario, num_envs=10, n_steps=10):
+    env = make_env(
+        scenario=scenario,
+        num_envs=num_envs,
+        seed=0,
+        multidiscrete_actions=True,
```
Comment (matteobettini) [inline tests/test_vmas.py:81]: Let's test both True and False here

File: tests/test_vmas.py [hunk 8]
```diff
@@ -70,6 +72,135 @@ def test_multi_discrete_actions(scenario, num_envs=10, n_steps=10):
         env.step(env.get_random_actions())
 
 
+@pytest.mark.parametrize("scenario", scenario_names())
+def test_discrete_action_nvec(scenario, num_envs=10, n_steps=10):
```
Comment (matteobettini) [inline tests/test_vmas.py:76]:
```suggestion
def test_discrete_action_nvec(scenario, num_envs=10, n_steps=5):
```

File: tests/test_vmas.py [hunk 9]
```diff
@@ -70,6 +72,135 @@ def test_multi_discrete_actions(scenario, num_envs=10, n_steps=10):
...
+        type(env.scenario).process_action
+        is not vmas.simulator.scenario.BaseScenario.process_action
+    ):
+        pytest.skip("Scenario uses a custom process_action method.")
+
+    random.seed(0)
+    for agent, agent_multi in zip(
+        env.world.policy_agents, env_multi.world.policy_agents
+    ):
+        nvec = [random.randint(2, 6) for _ in range(agent.action_size)]
```
Comment (matteobettini) [inline tests/test_vmas.py:154]: Should we test with different action sizes? casue most scnearios have `action_size=2`. For the same reason, testing this on all scenarios is not really needed as the logic is in the environment. One scenario with a parametric `action_size` and `nvec` would be enough.

File: tests/test_vmas.py [hunk 10]
```diff
@@ -70,6 +72,135 @@ def test_multi_discrete_actions(scenario, num_envs=10, n_steps=10):
...
+                            uj / k, 2 * U * (aj / (n - 1)) - U
+                        ), f"discrete action {aj} maps to control {uj} (n={n}), U={U}, k={k})"
+                    else:
+                        assert torch.isclose(
+                            uj / k, 2 * U * (aj / (n - 1)) - U
+                        ), f"discrete action {aj} maps to control {uj} (n={n}), U={U}, k={k})"
+
+
+@pytest.mark.parametrize("scenario", scenario_names())
+def test_discrete_action_nvec_discrete_to_multi(scenario, num_envs=10, n_steps=10):
```
Comment (matteobettini) [inline tests/test_vmas.py:135]:
```suggestion
def test_discrete_action_nvec_discrete_to_multi(scenario, num_envs=10, n_steps=5):
```

**Review by matteobettini — APPROVED** [review]
Stellar! 

Love the tests and can't wait to merge this.

Last few comments on tests
